### ISSUE TITLE
Spurious file-related failures on Windows runners

### ISSUE THREAD

**Issue by tgross35** [actions/runner-images#10483]
### Description

For the past few months, the `rust-lang/rust` project has had a lot of spurious failures on the Windows runners. These are typically either failure to open a file (mostly from `link.exe`) or failure to remove a file:

- `LINK : fatal error LNK1104: cannot open file ...`
- `error: failed to remove file ..., Access is denied (os error 5)`

Example run: https://github.com/rust-lang-ci/rust/actions/runs/10537107932/job/29198090275

These errors are similar to what was reported in https://github.com/actions/runner-images/issues/4086. Is it possible that something changed again? Even if not, we could use assistance debugging.

Further context, links to failed jobs, and attempts to reproduce are at https://github.com/rust-lang/rust/issues/127883.

Cc @ChrisDenton and @ehuss who have been working to reproduce this.

### Platforms affected

- [ ] Azure DevOps
- [ ] GitHub Actions - Standard Runners
- [X] GitHub Actions - Larger Runners

### Runner images affected

- [ ] Ubuntu 20.04
- [ ] Ubuntu 22.04
- [ ] Ubuntu 24.04
- [ ] macOS 12
- [ ] macOS 13
- [ ] macOS 13 Arm64
- [ ] macOS 14
- [ ] macOS 14 Arm64
- [ ] Windows Server 2019
- [X] Windows Server 2022

### Image version and build link

```text
Current runner version: '2.319.1'
Runner name: 'windows-2022-8core-32gb_4d2ba789d359'
Runner group name: 'Default Larger Runners'
Machine name: 'runner'
Operating System
  Microsoft Windows Server 2022
  10.0.20348
  Datacenter
Runner Image
  Image: windows-2022
  Version: 20240811.1.0
  Included Software: https://github.com/actions/runner-images/blob/win22/20240811.1/images/windows/Windows2022-Readme.md
  Image Release: https://github.com/actions/runner-images/releases/tag/win22%2F20240811.1
```

### Is it regression?

Yes, around 2024-06-27 but the exact start is unknown.

### Expected behavior

Accessing or removing the files should succeed.

### Actual behavior

The file operations are encountering spurious failures, as linked above.

### Repro steps

No known consistent reproduction.

**Comment by vidyasagarnimmagaddi** [comment]
Hi @tgross35 ,Thank you for bringing this issue to us. We are looking into this issue and will update you on this issue after investigating.

**Comment by vidyasagarnimmagaddi** [comment]
Hi  @tgross35  ,Thank you for bringing this issue to us. We are looking into this issue and will update you on this issue after investigating.

**Comment by tgross35** [comment]
Thank you for the response. If you need to watch active jobs there is always one running at https://github.com/rust-lang-ci/rust/actions (mostly the `auto` branch). `x86_64-msvc-ext` `dist-x86_64-msvc-alt` `dist-x86_64-msvc` all fail commonly, usually between 1 and 2 hours of job start.

**Comment by tgross35** [comment]
@vidyasagarnimmagaddi is there something we could do to debug this better? Our failure rate is currently over 50% due to this issue.

Somebody was able to confirm that we encounter this issue even running CI on an older state of our repo (from before this problem was noticed), which does seem to indicate it is caused by a change to the runner environment rather than changes to our code.

**Comment by ijunaidm** [comment]
@tgross35 - sure, we will update you shortly to provide workaround/solution to the issue.

**Comment by ehuss** [comment]
@ijunaidm Thanks! I'm one of the people working on this on the Rust side. Another datapoint: I've never been able to reproduce this on the `windows-2022` runner. I've only been able to do it on the `windows-2022-8core-32gb` runner. I don't know if that is just because of performance (`windows-2022` might be too slow to trip whatever race condition), or if it is fundamental to differences in the image. One thing I noticed is that `windows-2022-8core-32gb` uses the `C:` drive whereas `windows-2022` uses the `D:` drive. I'm not sure if that is relevant at all.

**Comment by tgross35** [comment]
@ijunaidm are there any updates here, or are you able to help us debug in some way (e.g. provide a way to ssh into active runners)? We were forced to switch to the small runners which seems to make this issue less prevalent (still very common) but need to move back to the large runners at some point.

**Comment by ijunaidm** [comment]
@tgross35 - Sorry, i will update you shortly on this issue.

**Comment by tgross35** [comment]
@subir0071 thank you for reaching out. 

> If possible exclude the deletion step from the workflow as for every new workflow run Github provisions new runners.

We create this file during the run and later need to delete it, so eliminating this step is not possible. Several people from rust-lang have already tried workarounds, including:

- Trying to rename the file if it fails to remove https://github.com/rust-lang/rust/pull/127152
- Retrying the delete if it fails https://github.com/rust-lang/rust/pull/127147
- Hacking around the way files are deleted https://github.com/rust-lang/rust/pull/129504
- Adding a 5 minute sleep before running the command that tries to delete https://github.com/rust-lang/rust/pull/129509
- Trying to figure out what is holding the file https://github.com/rust-lang/rust/pull/129836, https://github.com/rust-lang/rust/pull/129522
- Other misc attempts to figure out what is going on https://github.com/rust-lang/rust/pull/127373 https://github.com/rust-lang/rust/pull/129923

Further, this issue doesn't always take the same form; sometimes it is failure to remove a different file, sometimes it is failure to open files.

> Otherwise, make the pipeline wait to ensure the miri.exe is not busy, which can be another cause of access denied on deletion.

This is our main hypothesis for what is going on, but it doesn't seem to be us holding the file (meaning maybe there is some indexing, antivirus, or monitoring that was added within the past few months causing the problem). We have exhausted everything possible to figure out _why_ miri.exe would be busy. The PRs linked above have some attempts at this, see also e.g. https://github.com/rust-lang/rust/issues/127883#issuecomment-2336535344.

We cannot really do any further debugging here without GitHub's help, we need a way to do something like SSH into one of the runner images or create a kernel dump, or have somebody from GitHub do this for us while we run jobs. Is this possible?

**Comment by geekosaur** [comment]
@jasagredo, I think this will look familiar to you.

**Comment by jasagredo** [comment]
@geekosaur Not exactly, our issue was with permissions in read-only files that prevented Windows from deleting them. This seems to be related to Windows holding the lock for a file in use for too long.

@tgross35 
> SSH into one of the runner images

Maybe it is not very useful but there is https://github.com/mxschmitt/action-tmate which allows you to ssh into a runner, although I'm not sure of how elevated are the permissions there so you might not be able to dump protected stuff.

**Comment by tgross35** [comment]
@subir0071 I reverted our workarounds in https://github.com/rust-lang/rust/pull/134150 and started a test job. Usually the failures happen 1-2 hours after the job starts, I'll retry it if it doesn't fail.

Here is a link for the running jobs https://github.com/rust-lang-ci/rust/actions/runs/12266538500.

**Comment by tgross35** [comment]
@subir0071 it did fail with the same error that we had been experiencing before "failed to remove file" https://github.com/rust-lang/rust/pull/134150#issuecomment-2533392239

**Comment by tgross35** [comment]
@subir0071 thank you for looking into this. We would love to analyze this more on our side, but I need to reiterate that this error has never been reproduced on anything other than the GitHub Windows runners. We have tried debugging this but have reached the limits of what we can do within the limits of running jobs (`Handles` does not report anything), so we are stuck without help from GitHub.

Can you provide us with a way to SSH into runners so we can see what is using the files? Unfortunately https://github.com/actions/runner-images/issues/10483#issuecomment-2434913837 does not have sufficient permissions.

**Comment by ChrisDenton** [comment]
- The error message on delete is consistent with the application running while we try to delete it.
- However, `Handles` does not show it as being open and `procmon` shows the process exiting (and its file handle closed) long before the call to delete.
- We're experimenting with using a dev drive for builds, which so far appears to workaround the issue.
